File: src/lib/engine.js
import { AGE_COLUMNS, CONCERN_COLUMNS, QUIZ_OPTIONS, THRESHOLDS } from "./constants";
import { cleanText, loadProducts, normLabel } from "./data";
import { sanitizeProfile } from "./profiles";
import { concernAreaFor, matchesConcernArea } from "./concern-area";

function ageColumns(age) {
  return AGE_COLUMNS[normLabel(age)] || [];
}

function isSensitive(profile) {
  if (typeof profile.selectedSensitive === "boolean") return profile.selectedSensitive;
  const label = normLabel(profile.selectedSensitive);
  if (["yes", "true", "1", "y", "sensitive"].includes(label)) return true;
  if (["no", "false", "0", "n", "not sensitive", "none"].includes(label)) return false;
  return normLabel(profile.selectedSkinType).includes("sensitive");
}

function isUnder16(profile) {
  return JSON.stringify(ageColumns(profile.age)) === JSON.stringify(["<16"]);
}

function isDryOrSensitive(profile) {
  return normLabel(profile.selectedSkinType) === "dry" || isSensitive(profile);
}

function hasAntiAgingConcern(profile) {
  return [...(profile.selectedFaceBodyConcerns || []), ...(profile.selectedLipsEyesConcerns || [])]
    .some((concern) => {
      const label = normLabel(concern);
      return ["aging", "anti aging", "anti ageing", "wrinkles fine lines"].includes(label)
        || label.includes("wrinkle")
        || label.includes("fine line");
    });
}

function skinColumn(profile) {
  const base = ["Oily", "Dry", "Normal", "Combination"].includes(profile.selectedSkinType)
    ? profile.selectedSkinType
    : "Normal";
  return `${base}${isSensitive(profile) ? "+Sensitive" : ""} Score`;
}

export function normalizedProductType(value) {
  const type = cleanText(value).toLowerCase();
  if (["moisturiser", "moisturizer", "cream", "lotion"].includes(type)) return "moisturizer";
  if (["cleanser", "wash", "body wash"].includes(type)) return "cleanser";
  if (type === "seru") return "serum";
  return type;
}

function productRule(product, profile) {
  const type = normalizedProductType(product.product_type);
  if (type === "cleanser" && hasAntiAgingConcern(profile)) {
    return { age: false, concern: false, skin: true, special: true };
  }
  if (type === "serum") return { age: true, concern: true, skin: false, special: true };
  if (["moisturizer", "sunscreen"].includes(type)) {
    return { age: true, concern: false, skin: true, special: true };
  }
  return { age: true, concern: true, skin: true, special: true };
}

function displayScore(value) {
  return Number.isInteger(value) ? Math.trunc(value) : value;
}

function roundedAverage(components) {
  const scores = components.map(({ score }) => Number(score));
  if (!scores.length) return 0;
  if (scores.some((score) => score <= -100)) return -100;
  const average = scores.reduce((sum, score) => sum + score, 0) / scores.length;
  return average >= 0 ? Math.floor(average + 0.5) : Math.ceil(average - 0.5);
}

function firstScore(product, columns) {
  for (const column of columns) {
    if (product.scores[column] != null) return [column, product.scores[column]];
  }
  return null;
}

function drynessScore(value) {
  if ([-100, 0, 100].includes(value)) return value;
  if (value <= 50) return -100;
  if (value <= 84) return 0;
  return 100;
}

function specialScores(product, conditions, useDefaultNone = true) {
  const selected = conditions.map(normLabel).filter(Boolean);
  if (!selected.length || (selected.length === 1 && selected[0] === "none")) {
    return useDefaultNone ? [["None", "None", 100]] : [];
  }

  const matches = [];
  if (selected.includes("pregnant")) {
    const pair = firstScore(product, ["Pregnancy Score", "Pregnancy safe"]);
    if (pair) matches.push(["Pregnancy", pair[0], pair[1]]);
  }
  if (selected.includes("breastfeeding")) {
    const pair = firstScore(product, ["Breastfeeling Score", "Breastfeeding safe"]);
    if (pair) matches.push(["Breastfeeding", pair[0], pair[1]]);
  }
  if (selected.includes("excessive dryness")) {
    const pair = firstScore(product, ["Excessive Dryness score"]);
    if (pair) matches.push(["Excessive Dryness", pair[0], drynessScore(pair[1])]);
  }
  return matches;
}

function productText(product) {
  return [
    product.product_name,
    product.single_hero_ingredient,
    product.secondary_hero_ingredients,
    product.ingredients,
  ].join(" ").toLowerCase();
}

function isRetinoid(product) {
  const text = productText(product);
  return ["retinol", "retinal", "retinoid", "retinyl", "tretinoin", "adapalene"]
    .some((token) => text.includes(token));
}

function serumIsNightOnly(product, profile) {
  if (normalizedProductType(product.product_type) !== "serum") return false;
  const specials = (profile.selectedSpecialConditions || []).map(normLabel);
  return specials.some((item) => ["pregnant", "breastfeeding"].includes(item))
    || isUnder16(profile)
    || isDryOrSensitive(profile);
}

function matchLabel(score) {
  if (score >= 90) return "Excellent Match";
  if (score >= 80) return "Great Match";
  if (score >= 70) return "Good Match";
  if (score >= 50) return "Fits with Caution";
  return "Not Recommended";
}

function scoreProduct(product, profile) {
  const components = [];
  const reasons = [];
  const warnings = [];
  const rule = productRule(product, profile);

  const selectedAge = cleanText(profile.age);
  const agePairs = ageColumns(selectedAge)
    .filter((column) => product.scores[column] != null)
    .map((column) => [column, product.scores[column]]);
  if (rule.age && agePairs.length === 1) {
    const [column, score] = agePairs[0];
    components.push({ name: `Age ${column}`, score: displayScore(score), source_column: column });
    reasons.push(`Age fit ${column}: ${score}`);
  } else if (rule.age && agePairs.length > 1) {
    const score = roundedAverage(agePairs.map(([, value]) => ({ score: value })));
    components.push({
      name: `Age ${selectedAge}`,
      score,
      source_column: agePairs.map(([column]) => column).join(", "),
    });
    reasons.push(`Age fit ${selectedAge} via ${agePairs.map(([column, value]) => `${column}: ${value}`).join(", ")}`);
  }

  if (rule.concern) {
    const concerns = profile.selectedFaceBodyConcerns || [];
    for (const concern of concerns) {
      const columns = CONCERN_COLUMNS[normLabel(concern)] || [cleanText(concern)];
      const pair = firstScore(product, columns);
      if (pair) {
        components.push({ name: cleanText(concern), score: displayScore(pair[1]), source_column: pair[0] });
        reasons.push(`${cleanText(concern)} via ${pair[0]}: ${pair[1]}`);
      }
    }
    if (!concerns.length && product.scores.None != null) {
      components.push({ name: "General fit", score: displayScore(product.scores.None), source_column: "None" });
    }
  }

  if (rule.skin) {
    const column = skinColumn(profile);
    if (product.scores[column] != null) {
      components.push({ name: column, score: displayScore(product.scores[column]), source_column: column });
      reasons.push(`${column}: ${product.scores[column]}`);
    }
  }

  let specials = profile.selectedSpecialConditions || [];
  let useDefaultNone = true;
  if (normalizedProductType(product.product_type) === "serum") {
    const filtered = specials.filter((item) => normLabel(item) !== "excessive dryness");
    const skippedDryness = filtered.length !== specials.length;
    specials = filtered;
    useDefaultNone = !skippedDryness || filtered.map(normLabel).filter(Boolean).length > 0;
  }
  if (rule.special) {
    for (const [name, column, score] of specialScores(product, specials, useDefaultNone)) {
      components.push({ name, score: displayScore(score), source_column: column });
      reasons.push(`${name}: ${score}`);
      if (score < 0) warnings.push(`${name} has a hard-blocker score for this profile.`);
    }
  }

  const score = roundedAverage(components);
  const label = matchLabel(score);
  const hardBlockers = components.filter((component) => Number(component.score) <= -100);
  const used = components
    .slice(0, 6)
    .map((component) => `${component.name} [${component.source_column}]: ${component.score}`)
    .join("; ");
  let explanation;
  if (hardBlockers.length) {
    const blocked = hardBlockers
      .slice(0, 3)
      .map((component) => `${component.name} [${component.source_column}]: ${component.score}`)
      .join("; ");
    explanation = `${label}: this product has a hard blocker for this profile, so the final score is -100. Blocked by ${blocked}. Used ${used || "no matching score columns"}. Only products present in the live catalog CSV are returned.`;
    warnings.push("Hard blocker score found for this profile.");
  } else {
    const enabledRules = Object.entries(rule).filter(([, enabled]) => enabled).map(([name]) => name).join(", ");
    explanation = `${label}: score is the rounded average of applicable catalog scores for this profile. Used ${used || "no matching score columns"}. Product-type rule: ${cleanText(product.product_type) || "Unknown"} uses ${enabledRules} components. Only products present in the live catalog CSV are returned.`;
  }
  if (warnings.length) explanation = `${warnings[0]} ${explanation}`;

  return {
    product_uid: product.product_uid,
    score_uid: product.product_uid,
    product_name: product.product_name,
    brand_name: product.brand_name,
    category: product.category,
    product_type: product.product_type,
    score,
    match_label: label,
    explanation,
    source_sheet: product.source_sheet,
    hero_ingredient: product.single_hero_ingredient,
    secondary_hero_ingredients: product.secondary_hero_ingredients,
    size: product.sku_size,
    mrp: product.mrp,
    selling_price: product.sp,
    base_when_to_use: product.when_to_use,
    when_to_use: serumIsNightOnly(product, profile) || isRetinoid(product)
      ? "Night"
      : product.when_to_use || "Morning and Night",
    dos: product.dos,
    donts: product.donts,
    addresses_skin_concerns: product.addresses_skin_concerns,
    ingredient_cautions: product.ingredient_cautions,
    usage_instructions: product.usage_instructions,
    ingredients: product.ingredients,
    image: product.image,
    component_scores: components,
    score_basis: "product_type_rule_with_hard_blocker_or_rounded_average",
    score_reasons: reasons,
    warnings,
  };
}

function thresholdCounts(products) {
  return Object.fromEntries(
    THRESHOLDS.map((threshold) => [
      `above_${threshold}`,
      products.filter(({ score }) => score >= threshold).length,
    ]),
  );
}

function summarize(products) {
  const categories = {};
  const productTypes = {};
  for (const product of products) {
    categories[product.category] = (categories[product.category] || 0) + 1;
    productTypes[product.product_type] = (productTypes[product.product_type] || 0) + 1;
  }
  return {
    excellent_count: products.filter(({ score }) => score >= 90).length,
    great_or_better_count: products.filter(({ score }) => score >= 80).length,
    good_or_better_count: products.filter(({ score }) => score >= 70).length,
    caution_or_better_count: products.filter(({ score }) => score >= 50).length,
    threshold_counts: thresholdCounts(products),
    categories,
    product_types: productTypes,
    top_score: products[0]?.score ?? null,
    bottom_score: products.at(-1)?.score ?? null,
  };
}

function sortProducts(a, b) {
  if (a.score !== b.score) return b.score - a.score;
  if (a.category !== b.category) return a.category < b.category ? -1 : 1;
  const left = a.product_name.toLowerCase();
  const right = b.product_name.toLowerCase();
  return left === right ? 0 : left < right ? -1 : 1;
}

export async function recommend(input = {}, requestedLimit = 500) {
  const [profile, profileAdjustments] = sanitizeProfile(input);
  const catalog = await loadProducts();
  const eligible = profile.concernArea ? catalog.filter(product => matchesConcernArea(product, concernAreaFor(profile))) : catalog;
  const sorted = eligible.map((product) => {
    const scored = scoreProduct(product, profile);
    if (profileAdjustments.length) scored.warnings = [...profileAdjustments, ...scored.warnings];
    return scored;
  }).sort(sortProducts);
  const limit = Math.max(1, Math.min(Number(requestedLimit) || 500, 1000));
  const products = sorted.slice(0, limit);
  return {
    profile,
    input_profile: input,
    profile_adjustments: profileAdjustments,
    target_sheets: ["Face and body"],
    total_matches: sorted.length,
    returned: products.length,
    summary: summarize(products),
    products,
  };
}

export async function health() {
  const products = await loadProducts();
  return {
    data_source: "supabase:roopsee_products",
    catalog_products: products.length,
    score_rows: products.length,
    score_rows_by_sheet: { "Face and body": products.length },
    catalog_missing_score_count: products.filter((product) => !Object.keys(product.scores).length).length,
    catalog_missing_score: products.filter((product) => !Object.keys(product.scores).length).map((product) => product.product_uid),
    score_only_count: 0,
    score_only_uids: [],
  };
}

function coverageStatus(profile, summary) {
  const lipsEyeOnly = Boolean(profile.selectedLipsEyesConcerns?.length)
    && !profile.selectedFaceBodyConcerns?.length;
  if (lipsEyeOnly) {
    if (summary.good_or_better_count >= 4 && summary.great_or_better_count >= 2) return "Strong";
    if (summary.good_or_better_count >= 2) return "Limited but usable";
    return "Coverage gap";
  }
  if (summary.good_or_better_count >= 12 && summary.great_or_better_count >= 5) return "Strong";
  if (summary.good_or_better_count >= 8 && summary.great_or_better_count >= 3) return "Usable";
  return "Coverage gap";
}

export async function coverage(profiles, topN = 12) {
  const rows = await Promise.all(profiles.map(async (profile, index) => {
    const response = await recommend(profile, 1000);
    const counts = thresholdCounts(response.products);
    return {
      profile_id: `profile_${String(index + 1).padStart(3, "0")}`,
      profile,
      scoring_profile: response.profile,
      status: coverageStatus(response.profile, response.summary),
      total_matches: response.total_matches,
      excellent_count: response.summary.excellent_count,
      great_or_better_count: response.summary.great_or_better_count,
      good_or_better_count: response.summary.good_or_better_count,
      threshold_counts: counts,
      above_90: counts.above_90,
      above_80: counts.above_80,
      above_70: counts.above_70,
      above_60: counts.above_60,
      above_50: counts.above_50,
      top_products: response.products.slice(0, topN),
      target_sheets: response.target_sheets,
    };
  }));
  const statusCounts = {};
  for (const row of rows) statusCounts[row.status] = (statusCounts[row.status] || 0) + 1;
  const [products, healthSummary] = await Promise.all([loadProducts(), health()]);
  return {
    profile_count: rows.length,
    status_counts: statusCounts,
    catalog_products: products.length,
    catalog_missing_score_count: healthSummary.catalog_missing_score_count,
    rows,
  };
}

export async function optionsPayload(coverageModes) {
  return { quiz_options: QUIZ_OPTIONS, coverage_modes: coverageModes, health: await health() };
}

File: src/coverage_tool_files/roopsee_coverage/engine.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from .constants import THRESHOLDS
from .loaders import load_score_rows, parse_catalog
from .profile_rules import sanitize_profile
from .scoring import normalized_product_type, score_row_for_profile, summarize_results, target_sheets, threshold_counts
from .utils import norm_key

ROUTINE_SLOTS = {
    "am": [
        {"key": "cleanser", "label": "Cleanser", "aliases": {"cleanser", "wash"}},
        {"key": "moisturiser", "label": "Moisturiser", "aliases": {"moisturiser", "moisturizer"}},
        {"key": "sunscreen", "label": "Sunscreen", "aliases": {"sunscreen"}},
    ],
    "pm": [
        {"key": "serum", "label": "Serum", "aliases": {"serum"}},
        {"key": "moisturiser", "label": "Moisturiser", "aliases": {"moisturiser", "moisturizer"}},
        {"key": "cleanser", "label": "Cleanser", "aliases": {"cleanser", "wash"}},
    ],
}

WEEKLY_MASK_COUNT = 2

ROUTINE_TIERS = {
    "premium": {
        "label": "Premium",
        "description": "Score 90+ and effective price above Rs. 1000.",
        "min_score": 90,
        "min_price": 1000,
        "max_price": None,
    },
    "value_fit": {
        "label": "Value Fit",
        "description": "Best score with effective price below Rs. 1000.",
        "min_score": None,
        "min_price": None,
        "max_price": 1000,
    },
}


def normalize_product_type(product_type: str) -> str:
    return " ".join(str(product_type or "").strip().lower().split())


def product_search_text(product: dict[str, Any]) -> str:
    return " ".join(
        str(product.get(field, ""))
        for field in [
            "product_name",
            "product_type",
            "hero_ingredient",
            "single_hero_ingredient",
            "secondary_hero_ingredients",
            "ingredients",
        ]
    ).lower()


def price_amount(value: Any) -> float | None:
    text = str(value or "").strip()
    if not text:
        return None
    cleaned = "".join(char for char in text if char.isdigit() or char == ".")
    if not cleaned:
        return None
    try:
        return float(cleaned)
    except ValueError:
        return None


def effective_price(product: dict[str, Any]) -> float | None:
    return price_amount(product.get("selling_price")) or price_amount(product.get("mrp"))


def matches_routine_slot(product: dict[str, Any], slot: dict[str, Any]) -> bool:
    product_type = normalize_product_type(product.get("product_type", ""))
    normalized = normalized_product_type(product_type)
    text = product_search_text(product)
    key = slot["key"]
    if key == "cleanser":
        return normalized == "cleanser"
    if key == "serum":
        return normalized == "serum"
    if key == "moisturiser":
        return normalized == "moisturizer" or product_type in {"body lotion", "body cream"}
    if key == "sunscreen":
        return normalized == "sunscreen" or "sunscreen" in text or "spf" in text
    return product_type in slot["aliases"]


def is_mask_product(product: dict[str, Any]) -> bool:
    product_type = normalize_product_type(product.get("product_type", ""))
    text = product_search_text(product)
    return "mask" in product_type or "masque" in product_type or "mask" in text or "masque" in text


def matches_routine_tier(product: dict[str, Any], tier: dict[str, Any]) -> bool:
    price = effective_price(product)
    if price is None:
        return False
    min_score = tier.get("min_score")
    min_price = tier.get("min_price")
    max_price = tier.get("max_price")
    if min_score is not None and product["score"] < min_score:
        return False
    if min_price is not None and price <= min_price:
        return False
    if max_price is not None and price >= max_price:
        return False
    return True


def select_routine_product(products: list[dict[str, Any]], slot: dict[str, Any], tier: dict[str, Any]) -> dict[str, Any] | None:
    for product in products:
        if matches_routine_slot(product, slot) and matches_routine_tier(product, tier):
            return product
    return None


def select_best_mask_products(products: list[dict[str, Any]], count: int = WEEKLY_MASK_COUNT) -> list[dict[str, Any]]:
    selected: list[dict[str, Any]] = []
    seen_uids: set[str] = set()
    for product in products:
        uid = norm_key(product.get("product_uid") or product.get("product_name"))
        if uid in seen_uids or not is_mask_product(product):
            continue
        selected.append(product)
        seen_uids.add(uid)
        if len(selected) >= count:
            break
    return selected


def build_routine(products: list[dict[str, Any]]) -> dict[str, Any]:
    routine: dict[str, Any] = {
        "tiers": {},
        "missing_slots": [],
        "selection_basis": "highest_scored_product_per_routine_slot_from_current_profile_results_with_price_tiers_and_top_two_weekly_masks",
    }
    for tier_key, tier in ROUTINE_TIERS.items():
        tier_payload: dict[str, Any] = {
            "label": tier["label"],
            "description": tier["description"],
            "am": [],
            "pm": [],
        }
        for period, slots in ROUTINE_SLOTS.items():
            for slot in slots:
                product = select_routine_product(products, slot, tier)
                item = {
                    "tier": tier_key,
                    "period": period,
                    "slot": slot["key"],
                    "label": slot["label"],
                    "product": product,
                }
                tier_payload[period].append(item)
                if product is None:
                    routine["missing_slots"].append({
                        "tier": tier_key,
                        "period": period,
                        "slot": slot["key"],
                        "label": slot["label"],
                    })
        routine["tiers"][tier_key] = tier_payload

    routine["weekly"] = []
    weekly_products = select_best_mask_products(products)
    for index in range(WEEKLY_MASK_COUNT):
        product = weekly_products[index] if index < len(weekly_products) else None
        slot_key = f"mask_{index + 1}"
        label = f"Best Mask {index + 1}"
        item = {
            "period": "weekly",
            "slot": slot_key,
            "label": label,
            "product": product,
        }
        routine["weekly"].append(item)
        if product is None:
            routine["missing_slots"].append({"period": "weekly", "slot": slot_key, "label": label})

    routine["am"] = routine["tiers"].get("premium", {}).get("am", [])
    routine["pm"] = routine["tiers"].get("premium", {}).get("pm", [])
    return routine


class RecommendationEngine:
    def __init__(self, score_workbook: Path, products_csv: Path):
        self.score_workbook = score_workbook
        self.products_csv = products_csv
        self.catalog = parse_catalog(products_csv)
        self.score_rows = load_score_rows(score_workbook, products_csv)

        self.score_rows_by_catalog_key: dict[str, list[Any]] = {}
        self.score_only_uids: list[str] = []
        for row in self.score_rows:
            key = norm_key(row.product_uid)
            if key in self.catalog:
                self.score_rows_by_catalog_key.setdefault(key, []).append(row)
            else:
                self.score_only_uids.append(row.product_uid)

        self.catalog_missing_score = [
            product["product_uid"]
            for key, product in self.catalog.items()
            if key not in self.score_rows_by_catalog_key
        ]

    def recommend(self, profile: dict[str, Any], limit: int = 500) -> dict[str, Any]:
        scoring_profile, profile_adjustments = sanitize_profile(profile)
        wanted_sheets = target_sheets(scoring_profile)
        best_by_uid: dict[str, dict[str, Any]] = {}
        for key, rows in self.score_rows_by_catalog_key.items():
            catalog = self.catalog[key]
            for row in rows:
                if row.source_sheet not in wanted_sheets:
                    continue
                scored = score_row_for_profile(row, catalog, scoring_profile)
                if profile_adjustments:
                    scored["warnings"] = profile_adjustments + scored["warnings"]
                existing = best_by_uid.get(catalog["product_uid"])
                if existing is None or scored["score"] > existing["score"]:
                    best_by_uid[catalog["product_uid"]] = scored

        sorted_results = sorted(
            best_by_uid.values(),
            key=lambda item: (-item["score"], item["category"], item["product_name"].lower()),
        )
        results = sorted_results[: max(1, min(limit, 1000))]
        return {
            "profile": scoring_profile,
            "input_profile": profile,
            "profile_adjustments": profile_adjustments,
            "target_sheets": sorted(wanted_sheets),
            "total_matches": len(best_by_uid),
            "returned": len(results),
            "summary": summarize_results(results),
            "routine": build_routine(sorted_results),
            "products": results,
        }

    def routine(self, profile: dict[str, Any], limit: int = 1000) -> dict[str, Any]:
        response = self.recommend(profile, limit=limit)
        return {
            "profile": response["profile"],
            "input_profile": response["input_profile"],
            "profile_adjustments": response["profile_adjustments"],
            "target_sheets": response["target_sheets"],
            "total_matches": response["total_matches"],
            "returned": response["returned"],
            "routine": response["routine"],
        }

    def health(self) -> dict[str, Any]:
        by_sheet: dict[str, int] = {}
        for row in self.score_rows:
            by_sheet[row.source_sheet] = by_sheet.get(row.source_sheet, 0) + 1
        return {
            "score_workbook": str(self.score_workbook),
            "products_csv": str(self.products_csv),
            "catalog_products": len(self.catalog),
            "score_rows": len(self.score_rows),
            "score_rows_by_sheet": by_sheet,
            "catalog_missing_score_count": len(self.catalog_missing_score),
            "catalog_missing_score": self.catalog_missing_score,
            "score_only_count": len(self.score_only_uids),
            "score_only_uids": self.score_only_uids,
        }

    def coverage(self, profiles: list[dict[str, Any]], top_n: int = 12) -> dict[str, Any]:
        rows = self.coverage_rows(profiles, top_n=top_n)
        status_counts: dict[str, int] = {}
        for row in rows:
            status_counts[row["status"]] = status_counts.get(row["status"], 0) + 1

        return {
            "profile_count": len(rows),
            "status_counts": status_counts,
            "catalog_products": len(self.catalog),
            "catalog_missing_score_count": len(self.catalog_missing_score),
            "rows": rows,
        }

    def coverage_rows(self, profiles: list[dict[str, Any]], top_n: int = 0) -> list[dict[str, Any]]:
        rows = []
        for index, profile in enumerate(profiles, start=1):
            response = self.recommend(profile, limit=1000)
            summary = response["summary"]
            counts = threshold_counts(response["products"], THRESHOLDS)
            status = coverage_status(response["profile"], summary)
            rows.append({
                "profile_id": f"profile_{index:03d}",
                "profile": profile,
                "scoring_profile": response["profile"],
                "status": status,
                "total_matches": response["total_matches"],
                "excellent_count": summary["excellent_count"],
                "great_or_better_count": summary["great_or_better_count"],
                "good_or_better_count": summary["good_or_better_count"],
                "threshold_counts": counts,
                "above_90": counts["above_90"],
                "above_80": counts["above_80"],
                "above_70": counts["above_70"],
                "above_60": counts["above_60"],
                "above_50": counts["above_50"],
                "top_products": response["products"][:top_n],
                "target_sheets": response["target_sheets"],
            })
        return rows


def coverage_status(profile: dict[str, Any], summary: dict[str, Any]) -> str:
    lips_eye_only = bool(profile.get("selectedLipsEyesConcerns")) and not bool(profile.get("selectedFaceBodyConcerns"))
    if lips_eye_only:
        if summary["good_or_better_count"] >= 4 and summary["great_or_better_count"] >= 2:
            return "Strong"
        if summary["good_or_better_count"] >= 2:
            return "Limited but usable"
        return "Coverage gap"

    if summary["good_or_better_count"] >= 12 and summary["great_or_better_count"] >= 5:
        return "Strong"
    if summary["good_or_better_count"] >= 8 and summary["great_or_better_count"] >= 3:
        return "Usable"
    return "Coverage gap"
